Elliott/add secret api

elliott-davis opened the pull request:
![](https://media2.giphy.com/media/Dzmg5QAojwb6M/200w.gif)

thesentinels commented:
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

chefsalim commented on components/builder-depot/src/server.rs:
```diff
@@ -130,7 +137,12 @@ pub fn origin_create(req: &mut Request) -> IronResult<Response> {
+        Ok(origin) => {
+            // TED: If this fails we need to delete the origin and report an origin create failure
+            // currently there is no mechanism for deleting an origin
+            generate_origin_encryption_keys(origin.get_name(), req)?;
```
Hmm, yeah this is a bit tricky if there's no way for a user to recover.  
A secondary question is, how will we seed the already existing origins?

chefsalim replied:
If we are also seeding during the first download, then perhaps this generate call should be removed from here..

elliott-davis replied:
Yeah it's a bit of a pickle. I'm happy to write an internal only origin delete but I was afraid to do it as part of this PR. To answer the secondary question: If a user requests to upload a key and their origin doesn't have an existing encryption key, we generate one and the cli downloads it, then continues on as normal.

chefsalim replied:
OK yeah.. is there any use case that would not be satisfied if we just removed this generate call from here then?

elliott-davis replied:
It's probably safe to remove here

chefsalim commented on components/builder-depot/src/server.rs:
```diff
@@ -461,32 +473,34 @@ fn download_latest_origin_encryption_key(req: &mut Request) -> IronResult<Respon
+        Err(err) => {
+            if err.get_code() == ErrCode::ENTITY_NOT_FOUND {
+                generate_origin_encryption_keys(&origin, req)?;
+                return download_latest_origin_encryption_key(req);
```
Not a fan of a recursive call here - would be better to make this linear - generate_origin_encryption_keys could return the keys, not an IronResponse.

chefsalim commented on components/builder-depot/src/server.rs:
```diff
@@ -461,32 +473,34 @@ fn download_latest_origin_encryption_key(req: &mut Request) -> IronResult<Respon
-            if !check_origin_access(req, &origin).unwrap_or(false) {
-                return Ok(Response::with(status::Forbidden));
-            }
+fn generate_origin_encryption_keys(origin: &str, req: &mut Request) -> IronResult<Response> {
```
I'm not sure we even need a generation function at this level at all with the new strategy - this could all get collapsed inside of the get_origin_encryption_keys in originsrv (maybe renamed to get_or_create_origin_encrytpion_keys w/ a GetOrCreateOriginEncryptionKeys message).

chefsalim commented on components/builder-depot/src/server.rs:
```diff
@@ -2064,6 +2078,195 @@ pub fn download_latest_builder_key(req: &mut Request) -> IronResult<Response> {
+        _ => return Ok(Response::with(status::UnprocessableEntity)),
+    };
+
+    if !check_origin_access(req, origin).unwrap_or(false) {
```
You could move up the check and do it before even getting the body..

chefsalim replied:
Actually, validate_params already does  an origin check..

chefsalim commented on components/builder-depot/src/server.rs:
```diff
@@ -2064,6 +2078,195 @@ pub fn download_latest_builder_key(req: &mut Request) -> IronResult<Response> {
+    >(req, &db_priv_request) {
+        Ok(key) => {
+            let key_str = from_utf8(key.get_body()).unwrap();
+            debug!("KEY2: {:?}", key_str);
```
Maybe remove the debug, or change KEY2: to something more meaningful..

chefsalim commented on components/builder-depot/src/server.rs:
```diff
@@ -2064,6 +2078,195 @@ pub fn download_latest_builder_key(req: &mut Request) -> IronResult<Response> {
+}
+
+pub fn list_origin_secrets(req: &mut Request) -> IronResult<Response> {
+    let origin_name = match get_param(req, "origin") {
```
You could use validate_params here as well..

chefsalim commented on components/builder-originsrv/src/migrations/2018-03-05-224023_origin_secrets/down.sql:
```diff
@@ -0,0 +1 @@
+-- This file should undo anything in `up.sql`
```
Are we leaving these empty?  If so, would be good to add that prescriptive guidance to the migrations doc if we are not able to take advantage of the diesel rollbacks.

elliott-davis replied:
Yeah the rollbacks are sort of pointless for our usecase. I'll add a card to write up a doc on it

elliott-davis replied:
https://github.com/habitat-sh/builder/issues/275

chefsalim commented on components/builder-originsrv/src/migrations/2018-03-05-224023_origin_secrets/up.sql:
```diff
@@ -0,0 +1,53 @@
+          RETURNING *;
+      RETURN;
+    END
+$$ LANGUAGE plpgsql VOLATILE;
```
We wanted to standardize on SQL instead of plpgsql where possible - consider switching it over.

chefsalim commented on components/builder-originsrv/src/migrations/2018-03-05-224023_origin_secrets/up.sql:
```diff
@@ -0,0 +1,53 @@
+  BEGIN
+      DELETE FROM origin_secrets WHERE name = os_name AND origin_id = os_origin_id;
+  END
+$$ LANGUAGE plpgsql VOLATILE;
```
Same - convert to SQL

elliott-davis commented on components/builder-originsrv/src/migrations/2018-03-05-224023_origin_secrets/up.sql:
```diff
@@ -0,0 +1,53 @@
+  id bigint PRIMARY KEY DEFAULT next_id_v1('origin_secrets_id_seq'),
+  origin_id bigint REFERENCES origins(id),
+  owner_id bigint,
+  name varchar(256),
```
I think we should be using varchar in more places. Using text doesn't make sense for things like names.

chefsalim replied:
Hmm, why is that? It seems we're getting more flexibility with text, with basically same cost under the covers. Another thing that might be weird is comparison between text and varchars, looks like there are subtle differences in implicit comparison.  It might be better to just be consistent with the rest of the code for now... unless there's some other strong reason?

raskchanky commented on Cargo.toml:
```diff
@@ -18,6 +18,5 @@ members = [
   "components/github-api-client",
   "components/net",
-  "components/op",
-  "components/segment-api-client"
```
`segment-api-client` should remain in this repo.  I think @christophermaier just moved it over here the other day?

christophermaier replied:
Yes, please leave it!

elliott-davis replied:
oops - rebase fail

raskchanky commented on components/builder-depot/src/server.rs:
```diff
@@ -439,17 +446,15 @@ fn write_archive(filename: &PathBuf, body: &mut Body) -> Result<PackageArchive>
 
-    if !check_origin_access(req, &origin).unwrap_or(false) {
-        return Ok(Response::with(status::Forbidden));
-    }
```
Why was this check removed?

elliott-davis replied:
validate_params does the chef for us

raskchanky replied:
Oh, nice.  Good call.

raskchanky commented on components/builder-depot/src/server.rs:
```diff
@@ -461,32 +466,76 @@ fn download_latest_origin_encryption_key(req: &mut Request) -> IronResult<Respon
-                    }
-                }
-                Err(err) => Ok(render_net_error(&err)),
+    match helpers::get_origin(req, origin) {
```
Is `generate_origin_encryption_keys` only called in one place?  I couldn't tell definitively from this diff, but if the `download_latest_origin_encryption_key` function above is the only place that calls it, you can probably skip this call to `helpers::get_origin` (which I think makes a network trip to the db server, no?) since `download_latest_origin_encryption_key` already does that.

raskchanky commented on components/builder-depot/src/server.rs:
```diff
@@ -461,32 +466,76 @@ fn download_latest_origin_encryption_key(req: &mut Request) -> IronResult<Respon
+            private_key.set_origin_id(origin.get_id());
+
+            let pair = BoxKeyPair::generate_pair_for_origin(origin.get_name())
+                .expect("failed to generate origin encryption key pair");
```
Since the function we're in returns a `Result`, is it worth returning an `Err` here instead of calling `expect`?  I remember there was some conversation around the use of `expect` with the crypto functions, and I'm guessing that's how we ended up here.  It looks like the error module in this crate already has an enum for errors from the core crate, so it might work to just turn this into:

```rust
let pair = BoxKeyPair::generate_pair_for_origin(origin.get_name()).map_err(Error::HabitatCore)?;
```

If that doesn't work though, I wouldn't pursue it further.  It looked like an easy win, but maybe it's not.

raskchanky commented on components/builder-depot/src/server.rs:
```diff
@@ -461,32 +466,76 @@ fn download_latest_origin_encryption_key(req: &mut Request) -> IronResult<Respon
+            route_message::<OriginPrivateEncryptionKeyCreate, OriginPrivateEncryptionKey>(
+                req,
+                &private_request,
+            )?;
```
What happens if the `OriginPublicEncryptionKeyCreate` succeeds but the `OriginPrivateEncryptionKeyCreate` fails?  Is the user left in a bad state?  If they try to create a new key again, will it succeed if this was aborted in a partially finished state?

raskchanky commented on components/builder-depot/src/server.rs:
```diff
@@ -2064,6 +2113,186 @@ pub fn download_latest_builder_key(req: &mut Request) -> IronResult<Response> {
+                    status::UnprocessableEntity,
+                    "Missing value for field: `value`",
+                )));
+            }
```
I'm so stoked you're validating this kind of stuff.  I wish we had a better mechanism for DRYing this kind of thing up in our codebase.  (And just to be clear, I realize that adding one is outside of the scope of this PR 😁 )

raskchanky commented on components/builder-depot/src/server.rs:
```diff
@@ -2064,6 +2113,186 @@ pub fn download_latest_builder_key(req: &mut Request) -> IronResult<Response> {
+        }
+    };
+
+    debug!("Secret Metadata: {:?}", secret_metadata);
```
Will this make private keys show up in our logs?

elliott-davis replied:
No, it's just metadata about the encrypted payload. It will return the values here https://github.com/habitat-sh/core/blob/master/components/core/src/crypto/keys/box_key_pair.rs#L32-L37

raskchanky replied:
Ah, perfect.

raskchanky commented on components/builder-depot/src/server.rs:
```diff
@@ -2064,6 +2113,186 @@ pub fn download_latest_builder_key(req: &mut Request) -> IronResult<Response> {
+    match helpers::get_origin(req, origin) {
+        Ok(mut origin) => {
+            db_pub_request.set_owner_id(origin.get_owner_id());
+            db_pub_request.set_origin(origin.take_name());
```
Is it worth caching these values in local variables from the call to `helpers::get_origin` above to avoid another network round trip?

raskchanky commented on components/builder-depot/src/server.rs:
```diff
@@ -2064,6 +2113,186 @@ pub fn download_latest_builder_key(req: &mut Request) -> IronResult<Response> {
+    secret.set_name(body.name);
+    secret.set_value(body.value.clone());
+    match helpers::get_origin(req, &origin) {
+        Ok(origin) => secret.set_origin_id(origin.get_id()),
```
Some comment here re: caching values from the first call to `helpers::get_origin`.

raskchanky commented on components/builder-depot/src/server.rs:
```diff
@@ -2311,13 +2540,8 @@ where
-    );
     r.get(
-        "/origins/:origin/encryption_keys/latest",
+        "/origins/:origin/encryption_key",
```
It might be worth adding the `breaking-changes` label (do we have one yet?) because of this part - just for the sake of release notes.  I think the odds of anyone using our HTTP API directly (independent of the `hab` CLI) are low, but just in case, it's worth pointing out that this is a breaking change.

elliott-davis replied:
good point. I thought it was safe since we hadn't shipped to prod yet, but we did that yesterday

raskchanky commented on components/builder-originsrv/src/data_store.rs:
```diff
@@ -1813,6 +1825,83 @@ impl DataStore {
     }
+    pub fn create_origin_secret(
+        &self,
+        os: &originsrv::OriginSecretCreate,
```
Shouldn't this be `osc`?  :trollface: KIDDING KIDDING

elliott-davis replied:
:shaved_ice:

raskchanky commented on components/builder-originsrv/src/data_store.rs:
```diff
@@ -1813,6 +1825,83 @@ impl DataStore {
+        ).map_err(SrvError::OriginSecretCreate)?;
+        match rows.iter().nth(0) {
+            Some(row) => Ok(self.row_to_origin_secret(row)),
+            None => Err(SrvError::NoRowsReturnedAfterInsert()),
```
Wow, this is _fantastic_.

raskchanky commented on components/builder-originsrv/src/data_store.rs:
```diff
@@ -1813,6 +1825,83 @@ impl DataStore {
+            &[&(os_get.get_origin_id() as i64), &os_get.get_name()],
+        ).map_err(SrvError::OriginSecretGet)?;
+        if rows.len() != 0 {
+            // We just checked - we know there is a value here
```
All the comments justifying `unwrap()`s are one of my favorite parts of our code.  😆

raskchanky commented on components/builder-protocol/protocols/originsrv.proto:
```diff
@@ -413,6 +413,36 @@ message OriginProjectList {
+}
+
+message OriginSecretCreate {
+  optional OriginSecret secret = 1;
```
I ❤️ this pattern.

raskchanky commented on components/builder-protocol/src/originsrv.rs:
```diff
@@ -1158,6 +1158,66 @@ impl Serialize for OriginProjectList {
+    where
+        S: Serializer,
+    {
+        let mut state = serializer.serialize_struct("secret", 2)?;
```
I think the number in this call refers to the number of fields being serialized in the struct, so I _think_ this should be 4.  I'm assuming everything serializes normally though, so honestly, I'm a little baffled.

raskchanky commented:
This is great stuff.  I left some comments that might be worth thinking about but nothing that looked like a huge blocker.

elliott-davis commented:
@raskchanky @chefsalim I added the breaking change label to this since I changed an api endpoint. I doubt anyone is using the encryption keys feature since it's in the api only. Better safe than sorry.

raskchanky commented on components/builder-depot/src/server.rs:
```diff
@@ -439,55 +446,91 @@ fn write_archive(filename: &PathBuf, body: &mut Body) -> Result<PackageArchive>
+    route_message::<OriginPrivateEncryptionKeyCreate, OriginPrivateEncryptionKey>(
+        req,
+        &private_request,
+    )?;
```
So if I'm understanding the existing code correctly, if someone hits this function and the call to create the public key succeeds, and the call to create the private key fails for some transient reason, it's not a big deal.  We'll return an error to the user, but if the user tries again (assuming the transient issue has resolved), it will work and they'll be all set.  From what I could tell, the only unique constraint we have is on the full name of the key, which includes the revision, and the revision will change every time they generate a new key.  Do I have all that correct?

If so, then this seems fine as is.

elliott-davis replied:
yep that is all correct. I am thinking about revisiting the logic there to do an update on the public keys but that's a future iteration

raskchanky commented:
![](https://media.giphy.com/media/3ohs7GA9LpwWf5TtCw/giphy.gif)

elliott-davis commented:
@thesentinels approve

thesentinels commented:
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

thesentinels commented:
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/builder/builds/352462892).

thesentinels commented:
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/builder/builds/352462892).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

![gif-keyboard-3280869874741411265](https://cloud.githubusercontent.com/assets/4304/17755674/1e577b82-6490-11e6-96b7-1663d283c824.gif)
